### PR TITLE
feat: SFX Abi validates and recodes ingress from Substrate and Eth

### PR DESCRIPTION
**Pull request by MaciejBaj**
# Summary of changes

**Summarize** the description of your changes. 

If it fixes a bug or creates a feature, link the issue.

## Acceptance Checklist

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue):
    - [ ] Have you **written tests **to protect against future occurrences of the bug?
- [ ] New feature (non-breaking change which adds functionality)
    - [ ] Have you **seen it working** in a development environment?
    - [ ] Have you **written tests** for **happy** and **unhappy** paths?
    - [ ] Have you implemented this feature as per the **issue acceptance criteria**?
- [ ] Breaking change (a feature that would cause existing functionality not to work as expected
    - [ ] Is the breaking change **documented**?
    - [ ] Has any previous changes been **deprecated**?
- [ ] CI/CD
    - [ ] If introducing new actions, have you **looked at the action code** for anything spurious?
    - [ ] Have you written **custom scripts** for things that could be actions already on the marketplace?
    - [ ] If introducing new actions, have you **pegged a static version**?
- [ ] Documentation Update
    - [ ] Have you checked other documentation, such as the docs repository?
    - [ ] If updating script documentation, have you **tested it on both environments**?
- [ ] Substrate
    - [ ] RPC methods?
    - [ ] Chainspec, both JSON specs & the module?
    - [ ] Runtime versioning?
    - [ ] Benchmarks?
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

Feel free to explain in further detail your choices if this is a significant change.

### CODE REVIEW

File: types/src/to_filled_abi.rs [hunk 1]
```diff
@@ -0,0 +1,1105 @@
...
+    #[derive(Debug, PartialEq, Encode, Decode, RlpEncodable, RlpDecodable)]
+    struct Donation {
+        donor: Vec<u8>,
+        amount: u128,
+        donation_time: Option<u64>,
+        donation_version: u128,
+    }
+
+    #[derive(Debug, PartialEq, Encode, Decode, RlpEncodable, RlpDecodable)]
+    struct Campaign {
```
Comment (MaciejBaj) [inline types/src/to_filled_abi.rs:913]: with scale encoded Donation == encoded Campaign :)

File: types/src/to_abi.rs [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+            Abi::Value32(_) => 4,
+            Abi::Byte(_) => 1,
+            Abi::Bool(_) => 1,
+            // this needs to be multiplied by the length of the vec
+            Abi::Vec(_, field) => 1usize + field.get_size(),
+            Abi::Tuple(_, (field1, field2)) => field1.get_size() + field2.get_size(),
+        }
+    }
+
+    pub fn decode_topics_as_rlp(
```
Comment (AwesomeIbex) [inline types/src/to_abi.rs:94]: There is a lot of unwrapping here, and copying the data.

I would suggest taking the `input` as a `bytes::BytesMut` and stripping the slice as we take bits from it. Here we are cloning a fair amount.

**Review by AwesomeIbex — COMMENTED** [review]
Comment on improvements, happy to be extracted out to a reasonably high priority ticket as we shouldn't be copying the slice to decode.

File: types/src/to_abi.rs (hunk 2 above)
Comment (AwesomeIbex) [reply types/src/to_abi.rs:94]: See here for an example: https://github.com/t3rn/t3rn/blob/a665590c85e5d0c7b343bc52acd848469d6a50df/types/src/side_effect.rs#L217

**Review by AwesomeIbex — CHANGES_REQUESTED** [review]
Please remove leakages, happy for the rest to be ticketed as technical debt

File: types/abi/src/sfx_abi.rs [hunk 3]
```diff
@@ -160,23 +72,34 @@ impl SFXAbi {
...
+
+            // Check if arguments are equal after recoding to ordered_args_codec
+            let recoded_payload: Data =
+                filled_abi_matched_by_name.recode_as(payload_codec, ordered_args_codec)?;
+
+            match recoded_payload == *ordered_arg {
+                true => continue,
+                false =>
+                    return Err(DispatchError::Other(
+                        Box::leak(format!("SFXAbi::invalid payload argument for: '{current_arg_name_str:?}'; expected: {ordered_arg:?}; received and recoded: {recoded_payload:?}").into_boxed_str()),
```
[the remark this answers is not in the document]
Comment (MaciejBaj) [reply types/abi/src/sfx_abi.rs:130]: no leaking :<

File: primitives/src/portal.rs [hunk 4]
```diff
@@ -26,21 +28,59 @@ pub struct RegistrationData {
...
+        out_codec: Codec,
+    ) -> Result<Vec<u8>, DispatchError>;
+
+    fn verify_tx_inclusion(
+        gateway_id: [u8; 4],
+        submission_target_height: Vec<u8>,
+        encoded_inclusion_data: Vec<u8>,
+    ) -> Result<Vec<u8>, DispatchError>;
+
+    fn verify_event_inclusion_and_recode(
```
Comment (petscheit) [inline primitives/src/portal.rs:73]: 🔥

File: types/abi/src/standard.rs [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+    SFXAbi {
+        args_names: vec![
+            (b"asset_id".to_vec(), true),
+            (b"to".to_vec(), true),
+            (b"amount".to_vec(), true),
+        ],
+        ingress_abi_descriptors: PerCodecAbiDescriptors {
+            // assume all indexed in topics ("+")
+            for_rlp:
+                b"Assets:Log(asset_id+:Account20,from+:Account20,to+:Account20,amount+:Value128)"
```
Comment (petscheit) [inline types/abi/src/standard.rs:64]: In EVM chains, amounts are u256. We fine to just update it here?

**Review by petscheit — DISMISSED** [review]
🔥 🚒

File: primitives/Cargo.toml [hunk 6]
```diff
@@ -37,10 +37,13 @@ sp-finality-grandpa = { git = "https://github.com/paritytech/substrate", default
 sp-state-machine    = { git = "https://github.com/paritytech/substrate", branch = 'polkadot-v0.9.27', default-features = false }
 sp-trie             = { git = "https://github.com/paritytech/substrate", branch = 'polkadot-v0.9.27', default-features = false }
 
+pallet-grandpa-finality-verifier = { path = "../finality-verifiers/grandpa", default-features = false }
```
[the remark this answers is not in the document]
Comment (MaciejBaj) [reply primitives/Cargo.toml]: Removed

[229ee21](https://github.com/t3rn/t3rn/pull/667/commits/229ee211d7e927736293203148d8ee455b43973a)

File: types/abi/src/recode_scale.rs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                    no_strut_prefix_data = right;
+                    Ok(left.to_vec())
+                })
+                .collect::<Result<Vec<Vec<u8>>, DispatchError>>()?;
+
+        Ok((
+            chopped_field_data.into_iter(),
+            *memo_prefix
+                .first()
+                .expect("encoded_struct_chopper - Memo cannot be empty for structs"),
```
[the remark this answers is not in the document]
Comment (MaciejBaj) [reply types/abi/src/recode_scale.rs]: depreciated, no expect in latest implementation.

File: types/abi/src/recode.rs [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+        fields_iter_clone: IntoIter<Box<Abi>>,
+    ) -> Result<(FilledAbi, usize), DispatchError> {
+        match codec {
+            Codec::Scale => RecodeScale::event_to_filled(field_data, name, fields_iter_clone),
+            Codec::Rlp => RecodeRlp::event_to_filled(field_data, name, fields_iter_clone),
+        }
+    }
+}
+
+pub fn trim_bytes(input: &[u8], n: usize) -> &[u8] {
```
[the remark this answers is not in the document]
Comment (MaciejBaj) [reply types/abi/src/recode.rs]: Refactored ABI not to use any in-house method that mutates the bytes in the following commits:
- [refactor: rely on bytes crate vs split_bytes in ABI](https://github.com/t3rn/t3rn/pull/667/commits/07a3d418803ef3a179c21e9f363f1b9fd39bc96e)
- [refactor: remove the use of trim_bytes and rely on slicing in ABI](https://github.com/t3rn/t3rn/pull/667/commits/2881d54447692e174e99a1e55e2cefcc36c37ae7)
- [refactor: rewrite rlp topic decoding without take_last_n](https://github.com/t3rn/t3rn/pull/667/commits/09e242d5700b17b522763a6ab4f3cd924a696ef4)

File: types/src/to_abi.rs (hunk 2 above)
Comment (MaciejBaj) [reply types/src/to_abi.rs]: it's done, answered to following comment.

**Review by AwesomeIbex — DISMISSED** [review]

**Review by petscheit — APPROVED** [review]

**Pull request merged**